### ISSUE TITLE
cmd/compile: compile failed with "Value live at entry"

### ISSUE THREAD

**Issue by YueMengGu** [golang/go#53702]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.3 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/xxx/.cache/go-build"
GOENV="/Users/xxx/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/xxx/Coding/go/pkg/mod"
GONOPROXY="*.everphoto.cn,git.smartisan.com"
GONOSUMDB="*.everphoto.cn,git.smartisan.com"
GOOS="darwin"
GOPATH="/Users/xxx/Coding/go"
GOPRIVATE="*.everphoto.cn,git.smartisan.com"
GOPROXY="https://goproxy.cn,https://proxy.golang.org,direct"
GOROOT="/Users/xxx/Coding/github/go"
GOSUMDB="sum.golang.google.cn"
GOTMPDIR=""
GOTOOLDIR="/Users/xxx/Coding/github/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18.3"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/77/qcdftwkd6jb6945_b2k6n60h0000gp/T/go-build1272012810=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

```go
package main

type Elem struct {}
func  (*Elem) Wait(callback func()) {}

type Base struct {
  elem [8]*Elem
}

var g_val = 1
func (s *Base) Do() *int {
  resp := &g_val
  for _, e := range s.elem {
    e.Wait(func() {
      *resp = 0
    })
  }
  return resp
}

type Sub struct {
  *Base
}

func main() {
  a := Sub{}
  a.Do()
}
```

### What did you expect to see?

```shell
go build test.go
```

compile with no error

### What did you see instead?

> # command-line-arguments
> ./diversity_wrap.go:15:7: internal compiler error: 'Sub.Do.func1': Value live at entry. It shouldn't be. func Sub.Do.func1, node resp, value nil
> 
> Please file a bug report including a short program that triggers the error.
> https://go.dev/issue/new

it's ok with go1.16 and go1.17

I have tried patch which didn't work:
- https://go-review.googlesource.com/c/go/+/402854/
- https://go-review.googlesource.com/c/go/+/403995/

**Comment by cuonglm** [comment]
Writing `Base.Do` as:

```
func (s *Base) Do() *int {
	resp := &g_val
	for *resp != 0 {
		s.elem[0].Wait(func() { *resp = 0 })
	}
	return resp
}
```

is enough to trigger the bug.

**Comment by cuonglm** [comment]
git bisect points to https://go-review.googlesource.com/c/go/+/327871 cc @mdempsky 

Unified IR does not have this bug, because it generates its own wrapper instead of relying on `reflectdata`.

**Comment by mdempsky** [comment]
> git bisect points to https://go-review.googlesource.com/c/go/+/327871 cc @mdempsky

That's surprising. That commit should have only affect devirtualizationed, but the test case doesn't have any interface types.

If we revert just the typecheck.NeedITab stuff, does it fix the issue? (We shouldn't need it for unified IR, but maybe it was important to !unified mode.)

**Comment by cuonglm** [comment]
> That's surprising. That commit should have only affect devirtualizationed, but the test case doesn't have any interface types.

I think the problem comes from generating method wrapper, not interface types. After CL 327871, we now always do inlining for method wrapper after global escape analysis. Note that disable inlining will make it compiled in non-unified.

> If we revert just the typecheck.NeedITab stuff, does it fix the issue?

No.

**Comment by cuonglm** [comment]
This is now fixed at tip, could we close? @mdempsky

**Comment by mdempsky** [comment]
@cuonglm The issue is reportedly still a regression from Go 1.17. It's also not specifically a generics issue, so I think we should try to backport fixes for Go 1.18/1.19.

**Comment by cuonglm** [comment]
@gopherbot please backport this issue to go1.18 and go1.19
